test/evp_mac_test.c

```diff
@@ -0,0 +1,331 @@
+    0x99,0x8f,0x5f,0xc1
+};
+
+static const struct TEST_DATA {
+    int mac_id;
+    const unsigned char *in;
+    size_t in_len;
+    const unsigned char *key;
+    size_t key_len;
+    const unsigned char *expect;
+    size_t expect_len;
+    const char *cipher_name;
+    const char *md_name;
+    const unsigned char *iv;
+    size_t iv_len;
+} test_data[] = {
+    {
+        EVP_MAC_HMAC,
+        hmac_input, sizeof(hmac_input) - 1,
+        hmac_sha224_key, sizeof(hmac_sha224_key),
+        hmac_sha224_expected, sizeof(hmac_sha224_expected),
+        NULL, "SHA224",
+    },
+    {
+        EVP_MAC_CMAC,
+        cmac_input, sizeof(cmac_input),
+        cmac_key, sizeof(cmac_key),
+        cmac_expected, sizeof(cmac_expected),
+        "AES-128-CBC"
+    },
+    {
+        EVP_MAC_GMAC,
+        gmac_input, sizeof(gmac_input),
+        gmac_key, sizeof(gmac_key),
+        gmac_expected, sizeof(gmac_expected),
+        "AES-128-GCM", NULL,
+        gmac_iv, sizeof(gmac_iv),
+    },
+    {
+        EVP_MAC_KMAC128,
+        kmac128_input, sizeof(kmac128_input),
+        kmac128_key, sizeof(kmac128_key),
+        kmac128_expected, sizeof(kmac128_expected)
+    },
+    {
+        EVP_MAC_SIPHASH,
+        siphash_input, sizeof(siphash_input),
+        siphash_key, sizeof(siphash_key),
+        siphash_expected, sizeof(siphash_expected)
+    },
+    {
+        EVP_MAC_POLY1305,
+        poly1305_input, sizeof(poly1305_input),
+        poly1305_key, sizeof(poly1305_key),
+        poly1305_expected, sizeof(poly1305_expected)
+    },
+    {
+        EVP_MAC_BLAKE2S,
+        blake2s_input, sizeof(blake2s_input),
+        blake2s_key, sizeof(blake2s_key),
+        blake2s_expected, sizeof(blake2s_expected)
+    },
+    {
+        EVP_MAC_BLAKE2B,
+        blake2b_input, sizeof(blake2b_input),
+        blake2b_key, sizeof(blake2b_key),
+        blake2b_expected, sizeof(blake2b_expected)
+    }
+};
+
+static int do_test_copy_ctx(EVP_MAC_CTX *ctx, int mac_id,
+                            const unsigned char *in, size_t in_len,
+                            const unsigned char *expect, size_t expect_len)
+{
+    int ret = 0;
+    EVP_MAC_CTX *ctx_pre_init = NULL;
+    EVP_MAC_CTX *ctx_post_init = NULL;
+    EVP_MAC_CTX *ctx_post_update = NULL;
+    unsigned char out[EVP_MAX_MD_SIZE];
+    size_t len = 0;
+
+    if (!TEST_ptr(ctx_pre_init = EVP_MAC_CTX_new_id(mac_id))
+            || !TEST_ptr(ctx_post_init = EVP_MAC_CTX_new_id(mac_id))
+            || !TEST_ptr(ctx_post_update = EVP_MAC_CTX_new_id(mac_id)))
+        goto err;
+
+    if (!TEST_true(EVP_MAC_CTX_copy(ctx_pre_init, ctx))
+            || !TEST_true(EVP_MAC_init(ctx))
+            || !TEST_true(EVP_MAC_CTX_copy(ctx_post_init, ctx))
+            || !TEST_true(EVP_MAC_update(ctx, in, in_len))
+            || !TEST_true(EVP_MAC_CTX_copy(ctx_post_update, ctx))
+            || !TEST_true(EVP_MAC_final(ctx, out, &len)))
+        goto err;
+    if (!TEST_mem_eq(out, len, expect, expect_len))
```

Comment:
Merge with previous if?

Reply:
The grouping was deliberately done this way. What it really is trying to test is that it gives the same result regardless of where the copy is done (So the first block is the pre setup, and the second block is the actual test)

Reply:
I don't understand.  The the copy returns false, this line won't be reached, so you're only doing this test if the copy *is* done either way?

Reply:
In other words, these two snippets of pseudo-C are exactly equivalent:

``` C
if (!foo)
    goto err;
if (!bar)
    goto err;
```

``` C
if (!foo
    || !bar)
    goto err;
```

Reply:
You are correct. I can merge into one giant blob - it made more sense to me to break it up at the time (i.e initally the init/update/final being repeatedly called was actually giving different results on the final :))

Reply:
They are functionally equivalent, but convey a different creative expression (and as such are part of why code is even copyrightable in the first place).